sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NonAsyncTaskShouldNotReturnNull.cs

```diff
@@ -0,0 +1,93 @@
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class NonAsyncTaskShouldNotReturnNull : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4586";
+        private const string MessageFormat = "Do not return null from this method, instead return 'Task.FromResult<T>(null)', " +
+            "'Task.CompletedTask' or 'Task.Delay(0)'.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static ISet<KnownType> TaskTypes =
+            new HashSet<KnownType>
+            {
+                KnownType.System_Threading_Tasks_Task,
+                KnownType.System_Threading_Tasks_Task_T,
+            };
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var returnStatement = (ReturnStatementSyntax)c.Node;
+
+                    if (!returnStatement.Expression.IsKind(SyntaxKind.NullLiteralExpression))
+                    {
+                        return;
+                    }
+
+                    var methodDeclaration = returnStatement.FirstAncestorOrSelf<MethodDeclarationSyntax>();
+
+                    if (methodDeclaration != null &&
+                        !methodDeclaration.Modifiers.Any(m => m.IsKind(SyntaxKind.AsyncKeyword)) &&
+                        IsTaskReturnType(methodDeclaration, c.SemanticModel))
+                    {
+                        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, returnStatement.Expression.GetLocation()));
+                    }
+                },
+                SyntaxKind.ReturnStatement);
```

Comment:
I was tempted to look for method declaration to avoid reporting multiple times for the same method but as we don't handle secondary locations correctly in SonarLint/NuGet I decided to go this way.